File: src/pyramid.rs
use image::{imageops::FilterType, GenericImage, GenericImageView, Rgba, RgbaImage};
use log::{error, info};
use reqwest::{
    blocking::{multipart, Client},
    header::{HeaderMap, HeaderValue},
};
use std::{
    fs::{create_dir_all, read, File},
    io::copy,
    path::{Path, PathBuf},
    time::Instant,
};

use crate::utils::download_file;

const TILE_PIXEL_SIZE: u32 = 256;

pub fn pyramid_step(
    x: i32,
    y: i32,
    z: i32,
    base_zoom_level_tile_id: Option<String>,
    area_id: String,
    worker_id: &str,
    token: &str,
    base_api_url: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let tiles_dir_path = Path::new("tiles");

    if !tiles_dir_path.exists() {
        create_dir_all(tiles_dir_path)?;
    }

    let area_tiles_dir_path = tiles_dir_path.join(&area_id);

    if !area_tiles_dir_path.exists() {
        create_dir_all(&area_tiles_dir_path)?;
    }

    let client = Client::new();

    match base_zoom_level_tile_id {
        Some(tile_id) => {
            pyramid_step_base_zoom_level(
                &client,
                x,
                y,
                area_id,
                worker_id,
                token,
                base_api_url,
                &area_tiles_dir_path,
                tile_id,
            )?;
        }
        None => {
            pyramid_step_lower_zoom_level(
                &client,
                x,
                y,
                z,
                area_id,
                worker_id,
                token,
                base_api_url,
                &area_tiles_dir_path,
            )?;
        }
    }

    Ok(())
}

pub fn pyramid_step_base_zoom_level(
    client: &Client,
    x: i32,
    y: i32,
    area_id: String,
    worker_id: &str,
    token: &str,
    base_api_url: &str,
    area_tiles_dir_path: &PathBuf,
    tile_id: String,
) -> Result<(), Box<dyn std::error::Error>> {
    info!("Downloading the base high quality tile for tile {}", &tile_id);

    let start = Instant::now();

    let zoom_11_x_path = area_tiles_dir_path.join("11").join(x.to_string());

    if !zoom_11_x_path.exists() {
        create_dir_all(&zoom_11_x_path)?;
    }

    let zoom_11_tile_path = zoom_11_x_path.join(format!("{}.png", y));

    let zoom_11_tile_url = format!(
        "{}/api/map-generation/render-steps/{}/full-map",
        base_api_url, tile_id
    );

    let mut headers = HeaderMap::new();

    headers.append(
        "Authorization",
        HeaderValue::from_str(&format!("Bearer {}.{}", worker_id, token))?,
    );

    download_file(&client, &zoom_11_tile_url, &zoom_11_tile_path, Some(headers))?;

    let duration = start.elapsed();

    info!(
        "Base high quality tile for tile {} downloaded in {:.1?}",
        &tile_id, duration
    );

    info!(
        "Generating tiles for zoom 11, 12 and 13 for high quality tile {}",
        &tile_id
    );

    let start = Instant::now();

    let zoom_12_path = &area_tiles_dir_path.join("12");
    let zoom_12_x_path = &zoom_12_path.join((x * 2).to_string());
    let zoom_12_x_plus_1_path = &zoom_12_path.join((x * 2 + 1).to_string());

    if !zoom_12_x_path.exists() {
        create_dir_all(zoom_12_x_path)?;
    }

    if !zoom_12_x_plus_1_path.exists() {
        create_dir_all(zoom_12_x_plus_1_path)?;
    }

    let zoom_12_tiles_paths = [
        &zoom_12_x_path.join(format!("{}.png", (y * 2).to_string())),
        &zoom_12_x_plus_1_path.join(format!("{}.png", (y * 2).to_string())),
        &zoom_12_x_path.join(format!("{}.png", (y * 2 + 1).to_string())),
        &zoom_12_x_plus_1_path.join(format!("{}.png", (y * 2 + 1).to_string())),
    ];

    split_image_in_four(&zoom_11_tile_path, &zoom_12_tiles_paths)?;

    // (tile_path, file_name, form_part_name)
    let mut tiles_for_upload: Vec<(PathBuf, String, String)> = vec![];

    // Generate tiles for zoom 13
    let zoom_12_tiles = [
        [x * 2, y * 2],
        [x * 2 + 1, y * 2],
        [x * 2, y * 2 + 1],
        [x * 2 + 1, y * 2 + 1],
    ];

    for (i_12, [x_12, y_12]) in zoom_12_tiles.iter().enumerate() {
        let zoom_13_path = &area_tiles_dir_path.join("13");
        let zoom_13_x_path = &zoom_13_path.join((x_12 * 2).to_string());
        let zoom_13_x_plus_1_path = &zoom_13_path.join((x_12 * 2 + 1).to_string());

        if !zoom_13_x_path.exists() {
            create_dir_all(zoom_13_x_path)?;
        }

        if !zoom_13_x_plus_1_path.exists() {
            create_dir_all(zoom_13_x_plus_1_path)?;
        }

        let zoom_13_tiles_paths = [
            &zoom_13_x_path.join(format!("{}.png", (y_12 * 2).to_string())),
            &zoom_13_x_plus_1_path.join(format!("{}.png", (y_12 * 2).to_string())),
            &zoom_13_x_path.join(format!("{}.png", (y_12 * 2 + 1).to_string())),
            &zoom_13_x_plus_1_path.join(format!("{}.png", (y_12 * 2 + 1).to_string())),
        ];

        split_image_in_four(&zoom_12_tiles_paths[i_12], &zoom_13_tiles_paths)?;

        // Resize and upload zoom 13 tiles
        let mut i_13 = 0;

        let zoom_13_tiles = [
            [x_12 * 2, y_12 * 2],
            [x_12 * 2 + 1, y_12 * 2],
            [x_12 * 2, y_12 * 2 + 1],
            [x_12 * 2 + 1, y_12 * 2 + 1],
        ];

        for zoom_13_tile_path in zoom_13_tiles_paths {
            resize_image_in_place(zoom_13_tile_path, TILE_PIXEL_SIZE, TILE_PIXEL_SIZE)?;
            let [x_13, y_13] = zoom_13_tiles[i_13];

            tiles_for_upload.push((
                zoom_13_tile_path.clone(),
                format!("{}.png", y_13),
                format!("{}_{}_{}", 13, x_13, y_13),
            ));

            i_13 += 1;
        }
    }

    // Resize and upload zoom 12 tiles
    let mut i_12 = 0;

    for zoom_12_tile_path in zoom_12_tiles_paths {
        resize_image_in_place(zoom_12_tile_path, TILE_PIXEL_SIZE, TILE_PIXEL_SIZE)?;
        let [x_12, y_12] = zoom_12_tiles[i_12];

        tiles_for_upload.push((
            zoom_12_tile_path.clone(),
            format!("{}.png", y_12),
            format!("{}_{}_{}", 12, x_12, y_12),
        ));

        i_12 += 1;
    }

    // Resize and upload zoom 11 tile
    resize_image_in_place(&zoom_11_tile_path, TILE_PIXEL_SIZE, TILE_PIXEL_SIZE)?;

    tiles_for_upload.push((
        zoom_11_tile_path,
        format!("{}.png", y),
        format!("{}_{}_{}", 11, x, y),
    ));

    upload_base_zoom_tiles(
        &client,
        base_api_url,
        &area_id,
        worker_id,
        token,
        11,
        x,
        y,
        tiles_for_upload,
    )?;

    let duration = start.elapsed();

    info!(
        "Tiles for zoom 11, 12 and 13 for high quality tile {} generated in {:.1?}",
        &tile_id, duration
    );

    Ok(())
}

pub fn pyramid_step_lower_zoom_level(
    client: &Client,
    x: i32,
    y: i32,
    z: i32,
    area_id: String,
    worker_id: &str,
    token: &str,
    base_api_url: &str,
    area_tiles_dir_path: &PathBuf,
) -> Result<(), Box<dyn std::error::Error>> {
    info!("Zoom={} x={} y={}, Trying to download children tiles", z, x, y);

    let start = Instant::now();

    let children_tiles = [
        [x * 2, y * 2],
        [x * 2 + 1, y * 2],
        [x * 2, y * 2 + 1],
        [x * 2 + 1, y * 2 + 1],
    ];

    let mut child_images: [Option<image::DynamicImage>; 4] = [None, None, None, None];

    let mut headers = HeaderMap::new();

    headers.append(
        "Authorization",
        HeaderValue::from_str(&format!("Bearer {}.{}", worker_id, token))?,
    );

    for (i, [x_child, y_child]) in children_tiles.iter().enumerate() {
        let child_tile_url = format!(
            "{}/api/map-generation/pyramid-steps/{}/{}/{}/{}",
            base_api_url,
            area_id,
            z + 1,
            x_child,
            y_child
        );

        let child_tile_x_path = area_tiles_dir_path
            .join((z + 1).to_string())
            .join(&x_child.to_string());

        if !child_tile_x_path.exists() {
            create_dir_all(&child_tile_x_path)?;
        }

        let child_tile_path = child_tile_x_path.join(format!("{}.png", y_child));

        let mut response = client.get(&child_tile_url).headers(headers.clone()).send()?;

        if !response.status().is_success() && response.status().as_str() != "404" {
            error!(
                "Failed to download pyramide tile with url {}. Status: {}. Response: {:?}",
                response.status(),
                &child_tile_url,
                response.text()
            );

            return Err(Box::new(std::io::Error::new(
                std::io::ErrorKind::Other,
                "Failed to download file.",
            )));
        }

        let mut file = File::create(&child_tile_path)?;
        copy(&mut response, &mut file)?;

        let child_image = image::open(&child_tile_path).ok();
        child_images[i] = child_image;
    }

    let duration = start.elapsed();

    info!(
        "Zoom={} x={} y={}, children tiles (maybe) downloaded in {:.1?}",
        z, x, y, duration
    );

    info!("Zoom={} x={} y={}, merging and resizing children tiles", z, x, y);

    let start = Instant::now();

    // Merging children tiles
    let tile_x_path = area_tiles_dir_path.join(&z.to_string()).join(&x.to_string());

    if !tile_x_path.exists() {
        create_dir_all(&tile_x_path)?;
    }

    let mut tile_image = RgbaImage::from_pixel(TILE_PIXEL_SIZE * 2, TILE_PIXEL_SIZE * 2, Rgba([0, 0, 0, 0]));

    if let Some(image) = &child_images[0] {
        tile_image.copy_from(&image.to_rgba8(), 0, 0)?;
    }

    if let Some(image) = &child_images[1] {
        tile_image.copy_from(&image.to_rgba8(), TILE_PIXEL_SIZE, 0)?;
    }

    if let Some(image) = &child_images[2] {
        tile_image.copy_from(&image.to_rgba8(), 0, TILE_PIXEL_SIZE)?;
    }

    if let Some(image) = &child_images[3] {
        tile_image.copy_from(&image.to_rgba8(), TILE_PIXEL_SIZE, TILE_PIXEL_SIZE)?;
    }

    // Saving on disk and resizing
    let tile_path = tile_x_path.join(format!("{}.png", y));
    tile_image.save(&tile_path)?;
    resize_image_in_place(&tile_path, TILE_PIXEL_SIZE, TILE_PIXEL_SIZE)?;

    let duration = start.elapsed();

    info!(
        "Zoom={} x={} y={}, children tiles merged and resized in {:.1?}",
        z, x, y, duration
    );

    // Uploading tile
    upload_tile(
        &client,
        base_api_url,
        &tile_path,
        format!("{}.png", y),
        &area_id,
        z,
        x,
        y,
        worker_id,
        token,
    )?;

    Ok(())
}

/// Split an image in four parts: Top-left, Top-right, Bottom-left and Bottom-right
///
/// /// # Arguments
///
/// * `input_path` - The path of the image to be splitted in four.
/// * `output_paths` - An array of path where the resulting images should be writen.
///     [Top-left, Top-right, Bottom-left, Bottom-right]
///
fn split_image_in_four(
    input_path: &PathBuf,
    output_paths: &[&PathBuf; 4],
) -> Result<(), Box<dyn std::error::Error>> {
    // Load the input image
    let img = image::open(&Path::new(input_path))?;
    let (width, height) = img.dimensions();

    let half_width = width / 2;
    let half_height = height / 2;

    // Define regions and save each quarter
    let regions = [
        (0, 0, half_width, half_height),                    // Top-left
        (half_width, 0, half_width, half_height),           // Top-right
        (0, half_height, half_width, half_height),          // Bottom-left
        (half_width, half_height, half_width, half_height), // Bottom-right
    ];

    for (i, &(x, y, w, h)) in regions.iter().enumerate() {
        let sub_image = img.view(x, y, w, h).to_image(); // Extract sub-image
        sub_image
            .save(&output_paths[i])
            .expect("Failed to save output image");
    }

    Ok(())
}

fn resize_image_in_place(
    image_path: &PathBuf,
    width: u32,
    height: u32,
) -> Result<(), Box<dyn std::error::Error>> {
    let img = image::open(&Path::new(image_path))?;
    let resized_img = img.resize(width, height, FilterType::Lanczos3);
    resized_img.save(image_path)?;

    Ok(())
}

fn upload_tile(
    client: &Client,
    base_api_url: &str,
    file_path: &PathBuf,
    file_name: String,
    area_id: &str,
    zoom: i32,
    x: i32,
    y: i32,
    worker_id: &str,
    token: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    info!("Uploading tile zoom={} x={} y={}", zoom, x, y);
    let start = Instant::now();

    let file = read(file_path)?;

    let part = multipart::Part::bytes(file)
        .file_name(file_name)
        .mime_str("image/png")?;

    let form = multipart::Form::new().part("file", part);

    let url = format!(
        "{}/api/map-generation/pyramid-steps/{}/{}/{}/{}",
        base_api_url, area_id, zoom, x, y
    );

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}.{}", worker_id, token))
        .header("Origin", base_api_url)
        .multipart(form)
        .send()?;

    if response.status().is_success() {
        let duration = start.elapsed();

        info!("Tile zoom={} x={} y={} uploaded in {:.1?}", zoom, x, y, duration);
    } else {
        error!(
            "Failed to upload tile zoom={} x={} y={}: {} {}",
            zoom,
            x,
            y,
            response.status(),
            response.text()?
        );
    }

    Ok(())
}

fn upload_base_zoom_tiles(
    client: &Client,
    base_api_url: &str,
    area_id: &str,
    worker_id: &str,
    token: &str,
    zoom: i32,
    x: i32,
    y: i32,
    tiles: Vec<(PathBuf, String, String)>,
) -> Result<(), Box<dyn std::error::Error>> {
    info!("Uploading tiles for base level zoom={} x={} y={}", zoom, x, y);

    let start = Instant::now();

    let mut form = multipart::Form::new();

    for (tile_path, tile_file_name, tile_form_part_name) in tiles {
        let file = read(tile_path)?;

        let part = multipart::Part::bytes(file)
            .file_name(tile_file_name)
            .mime_str("image/png")?;

        form = form.part(tile_form_part_name, part);
    }

    let url = format!(
        "{}/api/map-generation/pyramid-steps/{}/base-level/{}/{}",
        base_api_url, area_id, x, y
    );

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}.{}", worker_id, token))
        .header("Origin", base_api_url)
        .multipart(form)
        .send()?;

    if response.status().is_success() {
        let duration = start.elapsed();

        info!(
            "Tiles for base level zoom={} x={} y={} uploaded in {:.1?}",
            zoom, x, y, duration
        );
    } else {
        error!(
            "Failed to upload tiles for base level zoom={} x={} y={}: {} {}",
            zoom,
            x,
            y,
            response.status(),
            response.text()?
        );
    }

    Ok(())
}

File: src/pyramid.ts
import { ensureDir } from "@std/fs/ensure-dir";
import { JobHandlingAdditionnalArguments } from "./models.ts";
import { PyramidJob } from "./next-job-schema.ts";
import {
  HIGH_QUALITY_TILE_PIXEL_SIZE,
  PYRAMID_STEP_DIR_NAME,
  PYRAMID_STEP_ENDPOINT_PATH,
  RENDER_STEP_ENDPOINT_PATH,
  TILE_PIXEL_SIZE,
} from "./constants.ts";
import { basename, join } from "@std/path";
import { exists } from "@std/fs/exists";
import { fetchWithRetryAndTimeout, log, removeIfExists } from "./utils.ts";
import sharp, { type OverlayOptions } from "sharp";

class NotFoundFetchError extends Error {}

export async function handlePyramidJob(
  { areaId, x, y, baseZoomLevelTileId, z }: PyramidJob["data"],
  options: JobHandlingAdditionnalArguments,
) {
  await ensureDir(PYRAMID_STEP_DIR_NAME);
  const areaTilesDirPath = join(PYRAMID_STEP_DIR_NAME, areaId);
  await ensureDir(areaTilesDirPath);

  if (baseZoomLevelTileId !== null) {
    await pyramidStepBaseZoomLevel(
      { areaId, areaTilesDirPath, xBase: x, yBase: y, tileId: baseZoomLevelTileId },
      options,
    );
  } else {
    await pyramidStepLowerZoomLevel(
      { areaId, areaTilesDirPath, zoom: z, x, y },
      options,
    );
  }
}

async function pyramidStepBaseZoomLevel(
  { areaId, areaTilesDirPath, xBase, yBase, tileId }: {
    areaId: string;
    areaTilesDirPath: string;
    xBase: number;
    yBase: number;
    tileId: string;
  },
  options: JobHandlingAdditionnalArguments,
) {
  const { mapantApiBaseUrl, mapantApiToken, mapantApiWorkerId, threadNumber } = options;

  const zoom12Tiles = getChildrenTiles({ zoom: 11, x: xBase, y: yBase });

  const allTiles = [
    { zoom: 11, x: xBase, y: yBase },
    ...zoom12Tiles,
    ...zoom12Tiles.flatMap(getChildrenTiles),
  ];

  try {
    log(`Tile zoom=11 x=${xBase} y=${yBase} | Generating pyramid tiles for zoom 11 and higher`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Downloading base tile`, {
      level: "info",
      threadNumber,
    });

    const zoom11XPath = join(areaTilesDirPath, "11", xBase.toString());
    await ensureDir(zoom11XPath);
    const zoom11TilePath = join(zoom11XPath, `${yBase}.png`);
    const zoom11TileUrl = `${mapantApiBaseUrl}${RENDER_STEP_ENDPOINT_PATH}/${tileId}/full-map`;
    await downloadPng(zoom11TileUrl, zoom11TilePath, options);

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Base tile downloaded`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Generating zoom 12 tiles`, {
      level: "info",
      threadNumber,
    });

    await createChildTiles({ zoom: 11, x: xBase, y: yBase, areaTilesDirPath });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Zoom 12 tiles generated`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Generating zoom 13 tiles`, {
      level: "info",
      threadNumber,
    });

    await Promise.all(
      zoom12Tiles.map(({ x, y }) => createChildTiles({ zoom: 12, x, y, areaTilesDirPath })),
    );

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Zoom 13 tiles generated`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | resizing all tiles`, {
      level: "info",
      threadNumber,
    });

    await Promise.all(allTiles.map((tile) => resizeTile({ ...tile, areaTilesDirPath })));

    log(`Tile zoom=11 x=${xBase} y=${yBase} | All tiles resized`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Uploading tiles`, {
      level: "info",
      threadNumber,
    });

    const formData = new FormData();

    await Promise.all(
      allTiles.map(async ({ zoom, x, y }) => {
        const file = await Deno.readFile(
          join(areaTilesDirPath, zoom.toString(), x.toString(), `${y}.png`),
        );

        formData.append(
          `${zoom}_${x}_${y}`,
          new Blob([file], { type: "image/png" }),
          `${y}.png`,
        );
      }),
    );

    await fetchWithRetryAndTimeout(
      `${mapantApiBaseUrl}${PYRAMID_STEP_ENDPOINT_PATH}/${areaId}/base-level/${xBase}/${yBase}`,
      {
        method: "POST",
        body: formData,
        headers: {
          "Origin": mapantApiBaseUrl,
          "Authorization": `Bearer ${mapantApiWorkerId}.${mapantApiToken}`,
        },
      },
    );

    log(`Tile zoom=11 x=${xBase} y=${yBase} | Tiles uploaded`, {
      level: "info",
      threadNumber,
    });
  } catch (e) {
    await Promise.allSettled(
      allTiles.map(({ zoom, x, y }) =>
        removeIfExists(join(areaTilesDirPath, zoom.toString(), x.toString(), `${y}.png`))
      ),
    );

    throw e;
  }
}

async function pyramidStepLowerZoomLevel(
  { areaId, areaTilesDirPath, zoom, x, y }: {
    areaId: string;
    areaTilesDirPath: string;
    zoom: number;
    x: number;
    y: number;
  },
  options: JobHandlingAdditionnalArguments,
) {
  const { mapantApiBaseUrl, mapantApiToken, mapantApiWorkerId, threadNumber } = options;

  log(`Tile zoom=${zoom} x=${x} y=${y} | Generating pyramid tile`, {
    level: "info",
    threadNumber,
  });

  log(`Tile zoom=${zoom} x=${x} y=${y} | Downloading children tiles`, {
    level: "info",
    threadNumber,
  });

  const baseUrl = `${mapantApiBaseUrl}${PYRAMID_STEP_ENDPOINT_PATH}/${areaId}/${zoom + 1}`;
  const xTilePath = join(areaTilesDirPath, zoom.toString(), x.toString());

  await Promise.all([
    ensureDir(xTilePath),
    ensureDir(join(areaTilesDirPath, (zoom + 1).toString(), (x * 2).toString())),
    ensureDir(join(areaTilesDirPath, (zoom + 1).toString(), (x * 2 + 1).toString())),
  ]);

  const tilePath = join(xTilePath, `${y}.png`);

  const childrenTiles = [
    { xChild: x * 2, yChild: y * 2 },
    { xChild: x * 2 + 1, yChild: y * 2 },
    { xChild: x * 2, yChild: y * 2 + 1 },
    { xChild: x * 2 + 1, yChild: y * 2 + 1 },
  ].map(({ xChild, yChild }) => ({
    url: `${baseUrl}/${xChild}/${yChild}`,
    path: join(areaTilesDirPath, (zoom + 1).toString(), xChild.toString(), `${yChild}.png`),
  }));

  try {
    const fetchChildrenTilesResults = await Promise
      .allSettled(
        childrenTiles.map(async ({ url, path }) => {
          await downloadPng(url, path, options);
          return path;
        }),
      );

    // Throw only if not 404 error
    for (const result of fetchChildrenTilesResults) {
      if (result.status === "rejected" && !(result.reason instanceof NotFoundFetchError)) {
        throw result.reason;
      }
    }

    log(`Tile zoom=${zoom} x=${x} y=${y} | Children tiles downloaded`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=${zoom} x=${x} y=${y} | Merging and resizing children tiles`, {
      level: "info",
      threadNumber,
    });

    const [topLeftResult, topRightResult, bottomLeftResult, bottomRightResult] =
      fetchChildrenTilesResults;

    const overlays: OverlayOptions[] = [];

    if (topLeftResult.status === "fulfilled") {
      overlays.push({ input: topLeftResult.value, top: 0, left: 0 });
    }
    if (topRightResult.status === "fulfilled") {
      overlays.push({ input: topRightResult.value, top: 0, left: TILE_PIXEL_SIZE });
    }
    if (bottomLeftResult.status === "fulfilled") {
      overlays.push({ input: bottomLeftResult.value, top: TILE_PIXEL_SIZE, left: 0 });
    }
    if (bottomRightResult.status === "fulfilled") {
      overlays.push({
        input: bottomRightResult.value,
        top: TILE_PIXEL_SIZE,
        left: TILE_PIXEL_SIZE,
      });
    }

    const buffer = await sharp({
      create: {
        width: TILE_PIXEL_SIZE * 2,
        height: TILE_PIXEL_SIZE * 2,
        channels: 4,
        background: { r: 255, g: 255, b: 255, alpha: 0 },
      },
    }).composite(overlays).png().toBuffer();

    await sharp(buffer).resize({ width: TILE_PIXEL_SIZE, height: TILE_PIXEL_SIZE }).toFile(
      tilePath,
    );

    log(`Tile zoom=${zoom} x=${x} y=${y} | Children tiles merged and resized`, {
      level: "info",
      threadNumber,
    });

    log(`Tile zoom=${zoom} x=${x} y=${y} | Uploading tile`, {
      level: "info",
      threadNumber,
    });

    const formData = new FormData();

    formData.append(
      "file",
      new Blob([await Deno.readFile(tilePath)], { type: "image/png" }),
      `${y}.png`,
    );

    await fetchWithRetryAndTimeout(
      `${mapantApiBaseUrl}${PYRAMID_STEP_ENDPOINT_PATH}/${areaId}/${zoom}/${x}/${y}`,
      {
        method: "POST",
        body: formData,
        headers: {
          "Origin": mapantApiBaseUrl,
          "Authorization": `Bearer ${mapantApiWorkerId}.${mapantApiToken}`,
        },
      },
    );

    log(`Tile zoom=${zoom} x=${x} y=${y} | Tile uploaded`, {
      level: "info",
      threadNumber,
    });
  } catch (e) {
    await removeIfExists(tilePath);

    throw e;
  }
}

async function resizeTile(
  { zoom, x, y, areaTilesDirPath }: {
    zoom: number;
    x: number;
    y: number;
    areaTilesDirPath: string;
  },
) {
  const path = join(areaTilesDirPath, zoom.toString(), x.toString(), `${y}.png`);

  const buffer = await sharp(path).resize({ width: TILE_PIXEL_SIZE, height: TILE_PIXEL_SIZE })
    .toBuffer();

  await sharp(buffer).toFile(path);
}

function getChildrenTiles({ zoom, x, y }: { zoom: number; x: number; y: number }) {
  return [
    { zoom: zoom + 1, x: x * 2, y: y * 2 },
    { zoom: zoom + 1, x: x * 2 + 1, y: y * 2 },
    { zoom: zoom + 1, x: x * 2, y: y * 2 + 1 },
    { zoom: zoom + 1, x: x * 2 + 1, y: y * 2 + 1 },
  ];
}

async function createChildTiles(
  { zoom, x, y, areaTilesDirPath }: {
    zoom: number;
    x: number;
    y: number;
    areaTilesDirPath: string;
  },
) {
  const parentTilePath = join(areaTilesDirPath, zoom.toString(), x.toString(), `${y}.png`);
  const upperZoomPath = join(areaTilesDirPath, (zoom + 1).toString());
  const upperZoomXPath = join(upperZoomPath, (x * 2).toString());
  await ensureDir(upperZoomXPath);
  const upperZoomXPlus1Path = join(upperZoomPath, (x * 2 + 1).toString());
  await ensureDir(upperZoomXPlus1Path);

  const metadata = await sharp(parentTilePath).metadata();
  if (
    metadata.width === undefined || metadata.height === undefined ||
    metadata.width !== metadata.height
  ) {
    throw new Error("Ban image format");
  }

  const side = Math.floor(metadata.width / 2);

  const extractOptions = [
    { left: 0, top: 0, path: join(upperZoomXPath, `${y * 2}.png`) },
    { left: side, top: 0, path: join(upperZoomXPlus1Path, `${y * 2}.png`) },
    { left: 0, top: side, path: join(upperZoomXPath, `${y * 2 + 1}.png`) },
    { left: side, top: side, path: join(upperZoomXPlus1Path, `${y * 2 + 1}.png`) },
  ];

  await Promise.all(
    extractOptions.map(({ left, top, path }) =>
      sharp(parentTilePath).extract({ left, top, width: side, height: side }).toFile(path)
    ),
  );
}

const downloadJobs: Map<string, Promise<void>> = new Map();

async function downloadPng(
  url: string,
  path: string,
  options: JobHandlingAdditionnalArguments,
): Promise<void> {
  const { threadNumber, mapantApiToken, mapantApiWorkerId } = options;
  const filename = basename(path);

  if (await exists(path)) {
    log(`File ${filename} already on disk`, { level: "info", threadNumber });
    return;
  }

  const onGoingJob = downloadJobs.get(url);

  if (onGoingJob !== undefined) {
    log(`File ${filename} already being downloaded`, { level: "info", threadNumber });
    return onGoingJob;
  }

  log(`Downloading file ${filename}`, { level: "info", threadNumber });

  const job = fetch(url, {
    headers: {
      "Authorization": `Bearer ${mapantApiWorkerId}.${mapantApiToken}`,
    },
  })
    .then(async (response) => {
      if (!response.ok || response.body === null) {
        const errorMessage =
          `Could not fetch file ${filename}, Status: ${response.status}, ${await response.text()}`;

        if (response.status === 404) throw new NotFoundFetchError();

        throw new Error(
          `Could not fetch file ${filename}, Status: ${response.status}, ${await response.text()}`,
        );
      }

      const file = await Deno.open(path, { write: true, create: true });
      return response.body.pipeTo(file.writable);
    })
    .then(() => {
      downloadJobs.delete(url);
    });

  downloadJobs.set(url, job);

  return job;
}
